### PR TITLE
Add filter_elements method on DocSet.

### PR DESCRIPTION
**Pull request by bsowell**
This is just a convenience method that behaves like map_elements.

### CODE REVIEW

File: lib/sycamore/sycamore/docset.py [hunk 1]
```diff
@@ -726,6 +726,21 @@ def custom_filter(doc: Document) -> bool:
...
+        """
+        Applies the given filter function to each element in each Document in this DocsSet.
+
+        Args:
+            f: A Callable that takes an Element and returns True if the element should be retained.
+        """
+
+        def process_doc(doc: Document) -> Document:
+            new_elements = filter(f, doc.elements)
+            doc.elements = new_elements
```
Comment (HenryL27) [inline lib/sycamore/sycamore/docset.py:739]: I've had some issues using `filter` with it giving me a 'filter' object which is an iterable but not a list (lazily evaluated)... mostly typing and serializability things. Might be bc I was trying to chain a bunch of them together, idk. Anyway, would suggest wrapping in `list( . )` or just writing out the comprehension

File: lib/sycamore/sycamore/tests/unit/test_docset.py [hunk 2]
```diff
@@ -199,3 +199,22 @@ def test_map_elements(self):
...
+            doc.elements = [
+                Element(text_representation=f"Document {i} Element {j}", properties={"element_val": j})
+                for j in range(10)
+            ]
+
+        context = sycamore.init()
+        docset = context.read.document(docs).filter_elements(lambda e: e.properties["element_val"] % 2 == 0)
+
+        all_docs = docset.take_all()
+        assert len(all_docs) == len(docs) / 2
```
Comment (HenryL27) [inline lib/sycamore/sycamore/tests/unit/test_docset.py:216]: wait there was no explode. shouldn't these be the same length? (but each doc only has 5 elts?)

Comment (HenryL27) [reply lib/sycamore/sycamore/tests/unit/test_docset.py:216]: oh it's bc you never appended the constructed `doc`s to `docs` so 0 = 0/2

Comment (bsowell) [reply lib/sycamore/sycamore/tests/unit/test_docset.py:216]: Oh oops. This means the test I copied this from was wrong too.

File: lib/sycamore/sycamore/docset.py (hunk 1 above)
Comment (bsowell) [reply lib/sycamore/sycamore/docset.py:739]: It's interesting that mypy didn't catch this, but I'll switch to a comprehension.

**Review by HenryL27 — APPROVED** [review]

**Pull request merged**